src/postgres_deparse.c

```diff
@@ -130,173 +94,583 @@ isOp(const char *val)
+	subpart->str = makeStringInfo();
+	subpart->indent = stmt->base_indent;
+	if (indent)
+		subpart->indent += state->opts.indent_size;
+	return subpart;
+}
+
+// Returns current active part, and initializes the first one (including a empty subpart) if needed
+static DeparseStatePart *
+deparseGetCurrentPart(DeparseState *state)
+{
+	DeparseStateStatement *stmt = state->current;
+	if (stmt->parts)
+		return (DeparseStatePart *) llast(stmt->parts);
+
+	DeparseStatePart *part = palloc0(sizeof(DeparseStatePart));
+	part->subparts = lappend(part->subparts, makeDeparseStateSubPart(state, stmt, part->indent_subparts));
+	stmt->parts = lappend(stmt->parts, part);
+	return part;
+}
+
+static StringInfo deparseGetCurrentStringInfo(DeparseState *state)
+{
+	DeparseStatePart *part = deparseGetCurrentPart(state);
+	DeparseStateSubPart *subpart = (DeparseStateSubPart *) llast(part->subparts);
+	Assert(subpart != NULL);
+	return subpart->str;
+}
+
+static void deparseAppendStringInfo(DeparseState *state, const char *fmt,...)
+{
+	StringInfo	str = deparseGetCurrentStringInfo(state);
+	int			save_errno = errno;
+
+	for (;;)
+	{
+		va_list		args;
+		int			needed;
+
+		/* Try to format the data. */
+		errno = save_errno;
+		va_start(args, fmt);
+		needed = appendStringInfoVA(str, fmt, args);
+		va_end(args);
+
+		if (needed == 0)
+			break;				/* success */
+
+		/* Increase the buffer size and try again. */
+		enlargeStringInfo(str, needed);
+	}
+}
+
+static void deparseAppendStringInfoString(DeparseState *state, const char *s)
+{
+	StringInfo str = deparseGetCurrentStringInfo(state);
+	appendStringInfoString(str, s);
+}
+
+static void deparseAppendStringInfoChar(DeparseState *state, char ch)
+{
+	StringInfo str = deparseGetCurrentStringInfo(state);
+	appendStringInfoChar(str, ch);
+}
+
+static void
+removeTrailingSpaceFromStr(StringInfo str)
+{
+	if (str->len >= 1 && str->data[str->len - 1] == ' ') {
+		str->len -= 1;
+		str->data[str->len] = '\0';
+	}
+}
+
+static void
+removeTrailingSpace(DeparseState *state)
+{
+	StringInfo str = deparseGetCurrentStringInfo(state);
+	removeTrailingSpaceFromStr(str);
+}
+
+static void
+deparseAppendSubpart(DeparseState *state)
+{
+	DeparseStatePart *part = deparseGetCurrentPart(state);
+
+	part->subparts = lappend(part->subparts, makeDeparseStateSubPart(state, state->current, part->indent_subparts));
+}
+
+static void
+deparseAppendBreakpoint(DeparseState *state)
+{
+	DeparseStatePart *part = deparseGetCurrentPart(state);
+
+	// Don't emit two consecutive breakpoints
+	if (((DeparseStateSubPart *) llast(part->subparts))->str->len == 0)
+		return;
+
+	deparseAppendSubpart(state);
+}
+
+static void
+deparseAppendCommaAndBreakpoint(DeparseState *state)
+{
+	if (state->opts.commas_start_of_line)
+	{
+		deparseAppendBreakpoint(state);
+		deparseAppendStringInfoString(state, ", ");
+	}
+	else
+	{
+		deparseAppendStringInfoChar(state, ',');
+		deparseAppendBreakpoint(state);
+	}
+}
+
+static void
```

Comment:
Add a comment that talks about what we mean with "major" keyword (since its not a term that is defined by Postgres / SQL Standard).

Reply:
Also, it would be good to rename this to stay more abstract, e.g. `deparseAppendPartGroup` (if we renamed part to part group).